include/power_grid_model/math_solver/iterative_pf_solver.hpp

```diff
@@ -0,0 +1,170 @@
+            max_dev = derived_solver.iterate_unknown(output.u);
+            sub_timer.stop();
+        }
+
+        // calculate math result
+        sub_timer = Timer(calculation_info, 2225, "Calculate Math Result");
+        calculate_result(y_bus, input, output);
+
+        // Manually stop timers to avoid "Max number of iterations" to be included in the timing.
+        sub_timer.stop();
+        main_timer.stop();
+
+        const auto key = Timer::make_key(2226, "Max number of iterations");
+        calculation_info[key] = std::max(calculation_info[key], (double)num_iter);
+
+        return output;
+    }
+
+    void calculate_result(YBus<sym> const& y_bus, PowerFlowInput<sym> const& input, MathOutput<sym>& output) {
+        // pending to correct
+        // call y bus
+        output.branch = y_bus.calculate_branch_flow(output.u);
+        output.shunt = y_bus.calculate_shunt_flow(output.u);
+
+        // prepare source and load gen
+        output.source.resize(source_bus_indptr_->back());
+        output.load_gen.resize(load_gen_bus_indptr_->back());
+        // loop all bus
+        for (Idx bus = 0; bus != n_bus_; ++bus) {
+            // source
+            for (Idx source = (*source_bus_indptr_)[bus]; source != (*source_bus_indptr_)[bus + 1]; ++source) {
+                ComplexValue<sym> const u_ref{input.source[source]};
+                ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
+                output.source[source].i = dot(y_ref, u_ref - output.u[bus]);
+                output.source[source].s = output.u[bus] * conj(output.source[source].i);
+            }
+
+            // load_gen
+            for (Idx load_gen = (*load_gen_bus_indptr_)[bus]; load_gen != (*load_gen_bus_indptr_)[bus + 1];
+                 ++load_gen) {
+                LoadGenType const type = (*load_gen_type_)[load_gen];
+                switch (type) {
+                    case LoadGenType::const_pq:
+                        // always same power
+                        output.load_gen[load_gen].s = input.s_injection[load_gen];
+                        break;
+                    case LoadGenType::const_y:
+                        // power is quadratic relation to voltage
+                        output.load_gen[load_gen].s =
+                            input.s_injection[load_gen] * cabs(output.u[bus]) * cabs(output.u[bus]);
+                        break;
+                    case LoadGenType::const_i:
+                        // power is linear relation to voltage
+                        output.load_gen[load_gen].s = input.s_injection[load_gen] * cabs(output.u[bus]);
+                        break;
+                    default:
+                        throw MissingCaseForEnumError("Power injection", type);
+                }
+                output.load_gen[load_gen].i = conj(output.load_gen[load_gen].s / output.u[bus]);
+            }
+        }
+    }
+
+   protected:
```

Comment:
you can make every data member private, because you make this class as a friend of DeriveSolver